### PR TITLE
Pass call05.f90

### PR DESCRIPTION
**Pull request by klausler**
The changes in this branch of pmk-call04 get the test call05.f90 to pass.  These changes also start enforcing the constraints on the use of assumed type parameters.

(To be clear: The branch pmk-calls has not yet been merged.  Branch pmk-call04 is a branch of pmk-calls, and this branch is a branch of pmk-call04.)

### CODE REVIEW

File: test/semantics/call05.f90 [hunk 1]
```diff
@@ -58,60 +62,72 @@ subroutine sup(x)
...
-    !ERROR: If a dummy or effective argument is unlimited polymorphic, both must be so
+    !ERROR: If a POINTER or ALLOCATABLE dummy or actual argument is polymorphic, both must be so
+    call spa(ma)
+    !ERROR: If a POINTER or ALLOCATABLE dummy or actual argument is unlimited polymorphic, both must be so
     call sup(pp)
-    !ERROR: If a dummy or effective argument is unlimited polymorphic, both must be so
+    !ERROR: If a POINTER or ALLOCATABLE dummy or actual argument is unlimited polymorphic, both must be so
     call sua(pa)
-    !ERROR: If a dummy or effective argument is unlimited polymorphic, both must be so
+    !ERROR: actual argument type 'CLASS(*)' is not compatible with dummy argument type 't'
```
Comment (psteinfeld) [inline test/semantics/call05.f90]: If you put a "The" into the message string in lib/evaluate/characteristics.cc, ```IsComptibleWith()```, we could guarantee that the error message would consistently be capitalized.

File: lib/semantics/check-declarations.cc [hunk 2]
```diff
@@ -76,6 +56,33 @@ class CheckHelper {
   const Scope *scope_{nullptr};
 };
 
+void CheckHelper::Check(const ParamValue &value, bool canBeAssumed) {
+  if (value.isAssumed()) {
+    if (!canBeAssumed) {  // C795
+      messages_.Say(
+          "An assumed (*) type parameter may be used only for a dummy argument, associate name, or named constant"_err_en_US);
```
[the remark this answers is not in the document]
Comment (klausler) [reply lib/semantics/check-declarations.cc]: My code is full of long message strings.  I will break them up later once the dust settles on wordings.

File: test/semantics/call05.f90 (hunk 1 above)
Comment (klausler) [reply test/semantics/call05.f90]: Yes, but that would not be amenable to translations into languages with fewer (Russian) or more (nearly all other Indo-European) than one form of their definite articles; I'm not sure it's worth doing.

File: lib/semantics/check-declarations.cc [hunk 3]
```diff
@@ -84,7 +91,18 @@ void CheckHelper::Check(Symbol &symbol) {
...
+    bool canHaveAssumedParameter{IsNamedConstant(symbol) ||
+        IsAssumedLengthCharacterFunction(symbol) ||
+        symbol.test(Symbol::Flag::ParentComp)};
+    if (const auto *object{symbol.detailsIf<ObjectEntityDetails>()}) {
+      canHaveAssumedParameter |= object->isDummy() ||
+          (object->isFuncResult() &&
+              type->category() == DeclTypeSpec::Character);
+    } else {
+      canHaveAssumedParameter |=
+          symbol.has<AssocEntityDetails>() || symbol.has<DerivedTypeDetails>();
```
[the remark this answers is not in the document]
Comment (klausler) [reply lib/semantics/check-declarations.cc]: Construct entities can have assumed type parameters in their type-specs.

I'll take out the reference to `DerivedTypeDetails`, thanks.

**Pull request merged**